### PR TITLE
Re-export more `rustc_span::symbol` things from `rustc_span`.

### PR DESCRIPTION
**Pull request by nnethercote**
`rustc_span::symbol` defines some things that are re-exported from `rustc_span`, such as `Symbol` and `sym`. But it doesn't re-export some closely related things such as `Ident` and `kw`. So you can do `use rustc_span::{Symbol, sym}` but you have to do `use rustc_span::symbol::{Ident, kw}`, which is inconsistent for no good reason.

This commit re-exports `Ident`, `kw`, and `MacroRulesNormalizedIdent`, and changes many `rustc_span::symbol::` qualifiers to `rustc_span::`. This is a 300+ net line of code reduction, mostly because many files with two `use rustc_span` items can be reduced to one.

r? @jieyouxu

### CODE REVIEW

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -68,7 +68,7 @@ mod span_encoding;
 pub use span_encoding::{DUMMY_SP, Span};
 
 pub mod symbol;
-pub use symbol::{Symbol, sym};
+pub use symbol::{Ident, MacroRulesNormalizedIdent, Symbol, kw, sym};
```
Comment (nnethercote) [inline compiler/rustc_span/src/lib.rs:71]: Sorry for the tedious review. This is the line of note. The rest is just shortening `rustc_span::symbol::` qualifiers.

**Comment by oli-obk** [comment]
I don't think I find this valuable. Reducing/merging imports is not very useful imo unless it allows removing of crate deps via reexports

**Comment by nnethercote** [comment]
I do think it's valuable. I think it's silly that you can import `Symbol` and `sym` via `rustc_span`, but the extremely similar names `Ident` and `kw` must be imported via `rustc_span::symbol`. It's one of those annoying rough edges and it shows up in a zillion files.

**Comment by jieyouxu** [comment]
I don't feel strongly either way since the nesting doesn't seem very steep. Anyway, I'll look at this tmrw since there's *several* files to go through.

**Comment by ytmimi** [comment]
@nnethercote I'd like to opt out of these changes for rustfmt.

File: compiler/rustc_lint/src/lints.rs [hunk 2]
```diff
@@ -2643,7 +2641,7 @@ impl<G: EmissionGuarantee> LintDiagnostic<'_, G> for ElidedNamedLifetime {
         //  but currently this lint's suggestions can conflict with those of `clippy::needless_lifetimes`:
         //  https://github.com/rust-lang/rust/pull/129840#issuecomment-2323349119
         // HACK: `'static` suggestions will never sonflict, emit only those for now.
-        if name != rustc_span::symbol::kw::StaticLifetime {
+        if name != rustc_span::kw::StaticLifetime {
```
Comment (jieyouxu) [inline compiler/rustc_lint/src/lints.rs:2644]: Question: can this just also import `kw` or is `kw` used as a local var here somewhere?

File: compiler/rustc_metadata/src/rmeta/decoder.rs [hunk 3]
```diff
@@ -874,7 +873,7 @@ impl MetadataBlob {
                         let def_key = root.tables.def_keys.get(blob, item).unwrap().decode(blob);
                         #[cfg_attr(not(bootstrap), allow(rustc::symbol_intern_string_literal))]
                         let def_name = if item == CRATE_DEF_INDEX {
-                            rustc_span::symbol::kw::Crate
+                            rustc_span::kw::Crate
```
Comment (jieyouxu) [inline compiler/rustc_metadata/src/rmeta/decoder.rs:876]: Question: can this just import `kw`?

File: src/librustdoc/clean/auto_trait.rs
Comment (jieyouxu) [inline src/librustdoc/clean/auto_trait.rs:1]: Question: @rust-lang/rustdoc would you like to opt-in to the import change, or do you want to keep this untouched?

File: src/tools/clippy/book/src/development/common_tools_writing_lints.md
Comment (jieyouxu) [inline src/tools/clippy/book/src/development/common_tools_writing_lints.md:1]: Question: @rust-lang/clippy do you want to opt-in to this import change?

File: src/tools/rustfmt/src/attr.rs
Comment (jieyouxu) [inline src/tools/rustfmt/src/attr.rs:1]: Problem: @nnethercote as @ytmimi requested, can you opt-out of the import changes for rustfmt?

**Review by jieyouxu — CHANGES_REQUESTED** [review]
Thanks for the PR. I looked at the changes individually:

- We should drop rustfmt changes as @ytmimi requested.
- We should ask T-rustdoc if they want to opt-in to the rustdoc import changes (or at least in *this* PR).
- We should ask T-clippy if they want to opt-in to the clippy import changes (or at least in *this* PR).
- Compiler changes seem reasonable to me. If we want to make the re-exports/imports of {`Ident`, `kw`, `sym`} by downstream compiler crates and tool users more consistent, I guess better get it done earlier than later.

**Comment by nnethercote** [comment]
The diff from this file sums up the motivation for this change:
```text
diff --git a/compiler/rustc_ast_lowering/src/format.rs b/compiler/rustc_ast_lowering/src/format.rs
index 653116e1fe0..22aa1e6fc20 100644
--- a/compiler/rustc_ast_lowering/src/format.rs
+++ b/compiler/rustc_ast_lowering/src/format.rs
@@ -6,8 +6,7 @@
 use rustc_data_structures::fx::FxIndexMap;
 use rustc_hir as hir;
 use rustc_session::config::FmtDebug;
-use rustc_span::symbol::{Ident, kw};
-use rustc_span::{Span, Symbol, sym};
+use rustc_span::{Ident, Span, Symbol, kw, sym};

 use super::LoweringContext;
```
It's silly that `Symbol` gets a shortcut import but the closely related `Ident` doesn't. Likewise, `sym` gets a shortcut but the closely related `kw` does not.

File: compiler/rustc_metadata/src/rmeta/decoder.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_metadata/src/rmeta/decoder.rs:876]: Good catch, it can.

Comment (nnethercote) [reply compiler/rustc_metadata/src/rmeta/decoder.rs:876]: :+1:

File: compiler/rustc_lint/src/lints.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_lint/src/lints.rs:2644]: :+1:

File: src/tools/rustfmt/src/attr.rs
Comment (nnethercote) [reply src/tools/rustfmt/src/attr.rs:1]: I've removed the rustfmt change. Ironically, rustfmt is the component with the least standard use of these names. It mostly does `use rustc_span::symbol` and then uses `symbol::Ident`.

**Comment by jieyouxu** [comment]
@nnethercote if you don't want to wait for T-clippy/T-rustdoc, we could also split this PR into compiler-only changes vs tools changes, where the tool changes can be sent to the relevant teams. I can r+ the compiler changes specifically.

File: src/librustdoc/clean/auto_trait.rs
Comment (nnethercote) [reply src/librustdoc/clean/auto_trait.rs:1]: I have removed the rustdoc changes.

File: src/tools/clippy/book/src/development/common_tools_writing_lints.md
Comment (nnethercote) [reply src/tools/clippy/book/src/development/common_tools_writing_lints.md:1]: I have removed the clippy changes.

File: src/tools/rustfmt/src/attr.rs
Comment (nnethercote) [reply src/tools/rustfmt/src/attr.rs:1]: done

**Comment by nnethercote** [comment]
I've trimmed this down to just the compiler changes and rebased. It's conflict-prone, so:

@bors r=jieyouxu p=1

**Pull request merged**